GVRf/Extensions/3DCursor/IODevices/gvrmobileiodevice/src/main/java/com/example/gvrmobileiodevice/GVRMobileIoDevice.java

```diff
@@ -0,0 +1,193 @@
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+/**
+ * Created by v.nidamarthy, d.knaan, p.mehta, r.rudhradevan on 8/26/16.
+ *
+ * Server - mobile with app, client - mobile/iodevice
+ * This class encodes the reading and writing of data to the app over Wifi,  //TODO adding bluetooth communication
+ * WifiDirectBroadcastReceiver establishes connection over WiFi
+ * Connect Server listens to the client and establishes connection and
+ * also the Input and Outputstreams
+ * ClientWrite writes data/feedback to the client
+ * ReadFromServer reads the incoming data i.e the orientation
+ *
+ */
+public class GVRMobileIoDevice extends IoDevice{
+    public static final String DEVICE_ID = "gvrmobile";
+    public static final int VENDOR_ID = 525;
+    public static final int PRODUCT_ID = 0;
+    public static final String VENDOR_NAME = "Samsung";
+    public static final String DEVICE_NAME = "GVR Mobile Device";
+    private static final String TAG = GVRMobileIoDevice.class.getSimpleName();
+    private final ServerInfo serverInfo;
+    private Context androidContext;
+
+    /* Fields for sensors and sockets */
+    private SensorManager mSensorManager;
+    private Sensor accelerometer;
+    private Sensor magnetometer;
+    private GVRContext mGVRContext;
+
+    /* Broadcast receiver side*/
+    private WifiP2pManager mManager;
+    private WifiP2pManager.Channel mChannel;
+    private IntentFilter mIntentFilter;
+    public final static int port = 8888;
+    private WifiDirectBroadcastReceiver mReceiver;
+
+    public enum Actions {ACTION_VIBRATE}   //enum to set feedback actions to the device
+
+    @Override
+    public void handleOutputEvent(OutputEvent event) {
+
+        try {
+            if (event.action == Actions.ACTION_VIBRATE.ordinal())
+            {
+                //Write a single bit to give haptic feedback, vibration to the client/mobile device
+                if(serverInfo.outputStream != null && serverInfo.client.getInetAddress().isReachable(100)) {
+                    new Thread(new ClientWrite(serverInfo)).start();
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    //ServerInfo class to handle the input and output streams and other flags
+    public class ServerInfo
+    {
+        public Socket client = null; //client acept socket
+        public ServerSocket serversocket; //serversocket
+        public InputStream inputStream;
+        public OutputStream outputStream;
+        public GVRMobileIoDevice gvrMobileIoDevice; //class reference to call functions from outside
+        int port;
+        boolean flag;
+        boolean position_mode;   //flag to change modes between position and rotation
+        public ServerInfo()      //TODO eventually will be using IMU to perform this function
+        {
+            port = 8888;   // port can be changed
+            flag = true;
+            position_mode = true;
```

Comment:
1. you can use an enum to define mode: POSITION, ROTATION. this might make it more readable compared to a flag. Also adding another mode is easier.